### ISSUE TITLE
[Accessibility]: Global `<Tooltip>` requirements

### ISSUE THREAD

**Issue by umpox** [sourcegraph/sourcegraph#32362]
### Problem description

Tooltips need to be reconsidered as part of WCAG 2.1.

See https://sarahmhigley.com/writing/tooltips-in-wcag-21/#:~:text=WCAG%202.1%20requires%20any%20content,will%20never%20overlap%20other%20content). for more information.

### Expected behavior

Acceptance criteria: see [this comment](https://github.com/sourcegraph/sourcegraph/issues/32362#issuecomment-1117506197) below.

**Comment by lrhacker** [comment]
After [some discussion in Slack](https://sourcegraph.slack.com/archives/C0HMGV90V/p1651002397547569) with @rrhyne and @quinnkeast, we determined there is design work needed to support the full set of requirements listed in this issue.

There is also possibly some more discussion needed to clarify what UX best meets both the WCAG 2.1 requirements and how we use tooltips in our applications. For example, do we need to audit the places we are using tooltips to determine if they need other changes to be fully accessible (thinking in the context of whether they show additional clarity versus additional information, per the examples in Slack)?

Also, to call out one part that I feel a little uncertain about: in [the example Codepen](https://codepen.io/smhigley/pen/KjoerX) linked by Sarah Higley's post, the tooltip will never reappear once it is dismissed - this matches the accessibility requirement, but I fear this could be rather confusing if it was accidentally dismissed and the user has no way to make it reappear (aside from reloading the entire page). Would it really be accessible to permanently remove content from the page with no direction on how to see it again? 🤔

**Comment by lrhacker** [comment]
It sounds like, rather than going with the original acceptance criteria that was based on [the guidance in this post](https://sarahmhigley.com/writing/tooltips-in-wcag-21/#:~:text=WCAG%202.1%20requires%20any%20content,will%20never%20overlap%20other%20content), we will do the following to meet [the WCAG success criterion 1.4.13](https://www.w3.org/WAI/WCAG21/Understanding/content-on-hover-or-focus.html):

1. **Dismissible**: Allow the tooltip to be dismissed using the ESC key, even if the tooltip was triggered via pointer hover.
    - There may be some additional technical considerations here for listening for ESC when the tooltip was triggered via pointer hover - can we reliably determine if ESC is meant to close the tooltip versus other content, such as a modal? This was a concern expressed in the original blog post referenced.
2. **Hoverable**: Make sure the tooltip content can also be hovered by a pointer to keep the content visible (rather than just the trigger).
    - With this, we want to make sure that moving the pointer from the trigger to the content does not require pixel-perfect precision.
4. **Persistent**: The tooltip should remain visible as long as the trigger maintains keyboard focus, or the pointer is hovered over the trigger or content, until focus/hover is changed or the ESC key is pressed.
    - The current implementation will dismiss the tooltip after cursor movement, even if the tooltip was triggered via keyboard focus, so we want to be sure we _don't_ carry that behavior over.

Based on these requirements, it seems like the [Tooltip from Adobe Spectrum](https://react-spectrum.adobe.com/react-spectrum/Tooltip.html) demonstrates all the behavior we want (with a 0s delay, in our case, since we want to show our tooltips immediately).

@rrhyne and @quinnkeast - does this accurately capture your interpretations of the success criterion and subsequent decisions?

**Comment by quinnkeast** [comment]
This sounds bang on to me!

**Comment by jasongornall** [comment]
Looks like yall arrived at the same conclusion but here is a good [example](https://www.w3.org/WAI/WCAG21/working-examples/script-hoverable/) to follow

**Comment by lrhacker** [comment]
#35870 resolved nearly all of these concerns. The one piece of functionality that is still outstanding is supporting focus events for elements that do not support them by default (for example, a plain text node). To fully support keyboard users, we need to make sure that all Tooltip trigger elements will capture focus events and show the Tooltip content when focused.

**Comment by lrhacker** [comment]
Due to issues we realized in #37321, the Tooltips no longer currently stay open on hover (they don't persist when the cursor is moved over them). Radix (the underlying library) is [working on support for this separately](https://github.com/radix-ui/primitives/issues/1475#issuecomment-1159051950), so it hopefully should be supported soon.

**Comment by umpox** [comment]
> The one piece of functionality that is still outstanding is supporting focus events for elements that do not support them by default (for example, a plain text node). To fully support keyboard users, we need to make sure that all Tooltip trigger elements will capture focus events and show the Tooltip content when focused.

Also, right now we support _any_ component as a tooltip trigger. This also causes some DX issues as we have to try to enforce that this components correctly use `forwardRef`, but we can't currently do that through TS/lint rules. 

A solution going forward might be to restrict tooltips to certain interactive components (Buttons, links, inputs). This will help:
- Fix accessibility issues, as keyboard users should definitely be able to trigger this components and thus see the tooltip
- Reduce the chance of users using an incorrect component as a tooltip trigger, improving DX

**Comment by felixfbecker** [comment]
Notes:
- Content can be hovered now, but it's a little fiddly, you need to hit the triangle
- Esc closes tooltip now, but if e.g. reference panel is open it will close that first.

**Comment by felixfbecker** [comment]
There is however another big problem with `<Tooltip>`, which is that it applies `aria-describedby` in a wrapper element, not to the element itself. This is an issue because it means screen readers don't actually announce it, because the focus will be on the child element (e.g. a button) not on the wrapper element. This is the cause of multiple other WCAG issues:
- https://github.com/sourcegraph/sourcegraph/issues/38782
- https://github.com/sourcegraph/sourcegraph/issues/41624
- https://github.com/sourcegraph/sourcegraph/issues/41683

**Comment by felixfbecker** [comment]
Closing in favor of remaining other issues.
